core/src/main/java/org/apache/calcite/tools/RelBuilder.java

```diff
@@ -2886,6 +2886,39 @@ private static boolean isGroupId(AggCall c) {
     return ((AggCallPlus) c).op().kind == SqlKind.GROUP_ID;
   }
 
+  /** Given a list of literals and a target row type, make the literals
+   * respectively match the fields types of the row.
+   *
+   * @param cluster  Cluster
+   * @param rowType  Type expected for values
+   * @param values   A list of literals that should match the rowType */
+  private static List<RexLiteral> convertLiteralTypes(RelOptCluster cluster,
+      RelDataType rowType, List<RexLiteral> values) {
+    assert values.size() == rowType.getFieldCount();
+    RexExecutor executor = cluster.getPlanner().getExecutor();
+    if (executor == null) {
+      executor = RexUtil.EXECUTOR;
+    }
+    if (executor == null) {
+      return values;
+    }
+
+    List<RelDataTypeField> fields = rowType.getFieldList();
+    List<RexLiteral> constants = new ArrayList<>();
+    RexBuilder builder = cluster.getRexBuilder();
+    RexSimplify simplify = new RexSimplify(builder, RelOptPredicateList.EMPTY, executor);
```

Comment:
Nit: Could the RexSimplify be passed as a parameter of this method? That would make the code simpler & cleaner, and we would avoid (re)creating a RexSimplify on each iteration that calls `convertLiteralTypes`.